### DISCUSSION TITLE
In nested models, accessing the data in the parent model and passing values to the sub-model?

### DISCUSSION THREAD

**Discussion by mhmtkcm** [pydantic/pydantic#7298] [Question]
Hello everyone,

I have some questions about nested models in Pydantic 2. I would appreciate if you could answer them.

Let's say I have nested models like below:

```python
class Title(BaseModel):
    name: str
    description: str

class Project(BaseModel):
    tools: list[str]
    type: str
    is_active: bool
    title: Title

class Experience(BaseModel):
    company_name: str
    start_date: str
    end_date: str
    titles: list[Title]
    projects: list[Project]

class Person(BaseModel):
    name: str
    age: int
    experinces: list[Experience]
```

**Questions:**

1. How can I exclude the **titles** field in the **Experience** model during validation by reading the age value in the **Person** model? For example, if the age is less than 25, there will be no titles field in the incoming data, so I need to specify that this field should not be expected.
2. If Person's age is greater than 45, let's say that the data in the **tools** field in the **Projects** model must contain the Python value. How do I pass the data in the main model to the sub-model of the sub-model?
3. In the models I gave in the example, it can be seen that the **Title** model is used as a sub-model in 2 different models. Can I determine which parent model the data in the **Title** field comes from and make a validation accordingly? For example, if the data comes from the **Project** model, I want to exclude the **description** field, but if it comes from the **Experince** model, I don't want to do anything.

In summary, the topics I want to learn in the nested model structure are as follows:
- Accessing a data in the parent model from the child model or from a model below the child model
- Exclude a field in a child model based on a validated data in the parent model
- When using common sub-models, to determine from which parent model the data comes to the sub-model and to process accordingly in sub-model

_I want to solve all the issues I asked above in the sub-model if possible. Therefore, it is not a solution for me to send all the data to the sub-model after validating it in the main model. (When determining the fields to be excluded, it may be necessary to solve it in the upper model, I have no idea about that)_

Thank you.

**Comment by aakashnarang1990** [comment]
did you find any solution to this ?

**Comment by mhmtkcm** [comment]
> did you find any solution to this ?

Unfortunately, not yet.

I am still looking for a solution.

**Comment by bartdw** [comment]
I also strungled with this for a long time. This is how I solved it. It meets your first 2 questions, not the 3'rd.
I hope it's useful for someone.

```python
from pydantic import BaseModel, Field, field_validator
from typing import Any, List, Union

class MyBaseModel(BaseModel):
    parent: Any = Field(default=None, exclude=True)

    @field_validator('*', mode='before')
    @classmethod
    def add_parent(cls, value, info) -> str:
        if info.field_name != 'parent' and type(value) is dict:
            value['parent'] = info.data
        if info.field_name != 'parent' and type(value) is list:
            for i in value:
                if type(i) is dict:
                    i['parent'] = info.data
        return value


class Title(MyBaseModel):
    name: str
    description: str

class Project(MyBaseModel):
    tools: list[str]
    type: str
    is_active: bool
    title: Title

    @field_validator('tools')
    @classmethod
    def expect_python_when_over_45(cls, value, info) -> str:
        age = info.data.get('parent').get('parent').get('age')
        if age > 45 and 'Python' not in value:
            raise ValueError('You should know Python if older then 45')
        return value

class Experience(MyBaseModel):
    company_name: str
    start_date: str
    end_date: str
    titles: Union[List[Title], None] = Field(default=None, validate_default=True)
    projects: list[Project]

    @field_validator('titles')
    @classmethod
    def expect_titles_when_under_25(cls, value, info) -> str:
        age = info.data.get('parent').get('age')
        if age > 25 and value is None:
            raise ValueError('Experience titles required when older then 25')
        return value

class Person(MyBaseModel):
    name: str
    age: int
    experiences: List[Experience]

data = {
    'name': 'foo',
    'age': 46,
    'experiences':[
        {
            'company_name': 'google',
            'start_date': 'yesterday',
            'end_date': 'tomorrow',
            'titles': [
                {
                    'name': 'engineer',
                    'description': 'blablah'
                }
            ],
            'projects': [
                {
                    'tools': ['git', 'Python'],
                    'type': 'something',
                    'is_active': False,
                    'title': {
                        'name': 'some_project',
                        'description': 'blablah'
                    }
                }
            ]
        }
    ]
}
m = Person.model_validate(data)
print(m.model_dump_json(indent=2))

del(data['experiences'][0]['titles'])
m = Person.model_validate(data)
print(m.model_dump_json(indent=2))
```

**Comment by LuchiLucs** [comment]
I am facing the same design problem. I would more inputs to the problem if other users solved it

**Comment by lionpeloux** [comment]
Maybe this can help : https://docs.pydantic.dev/latest/examples/custom_validators/#validating-nested-model-fields
